peps/pep-0807.rst

```diff
@@ -0,0 +1,398 @@
+non-PyPI implementations of Trusted Publishing.
+
+Specification
+=============
+
+This PEP's specification contains two parts:
+
+* A *discovery* mechanism that package upload clients can use to determine
+  whether an arbitrary Python package index host supports Trusted Publishing.
+* A *token exchange* mechanism that package upload clients can use to
+  exchange an identity credential for an upload credential.
+
+.. _constraints:
+
+Constraints
+-----------
+
+Unless explicitly stated otherwise, the following constraints
+apply to all parts of this PEP's specification:
+
+* All URLs **MUST** have
+  `potentially trustworthy origins <https://www.w3.org/TR/secure-contexts/#potentially-trustworthy-origins>`_.
+  In practice, this means that all URLs **MUST** use the ``https``
+  scheme, be some variant of a local loopback (``localhost``,
+  ``127.0.0.1``, etc.), or otherwise be considered *a priori* trustworthy
+  in the context of the interaction (e.g. an internal network).
+
+  Uploading clients **MUST** reject any URLs that do not meet this constraint.
+
+* All server-supplied URLs (i.e. those in discovery responses) **MUST**
+  have the same host subcomponent as the user-provided upload URL. Uploading
+  clients **MUST** reject any URLs that do not meet this constraint.
+
+  In practice, this means that a discovery request to
+  ``https://upload.example.com/.well-known/pytp/{key}`` can only
+  return URLs with the ``upload.example.com`` host.
+
+* All client requests **SHOULD** have an
+  ``Accept: application/vnd.pypi.pytp.v1+json`` header. In the absence of
+  an ``Accept`` header, the receiving server **MUST** behave as if this header
+  were present.
+
+  Receiving servers **SHOULD** respond with a ``406 Not Acceptable``
+  status code if any other ``Accept`` header is present.
+
+.. _discovery:
+
+Trusted Publishing Discovery
+----------------------------
+
+All Python package uploading is currently "endpoint driven," in the sense
+uploading clients (like twine and uv) are given an upload URL (and
+**not** merely a domain name).
+
+For example, to upload to PyPI, uploading clients are expected to connect
+to ``https://upload.pypi.org/legacy/``.
+
+The discovery mechanism proposed below takes advantage of this fact to
+allow single domains to advertise support for multiple indices
+(and their corresponding upload endpoints).
+
+The discovery mechanism is as follows:
+
+1. The uploading client is given an upload URL, e.g.
+   ``https://upload.example.com/legacy/``.
+
+2. The uploading client extracts the *path component* of the URL,
+   as defined in :rfc:`3986`. If the path component is empty,
+   the empty string should be used.
+
+   For the above example, the path component is
+   ``/legacy/``.
+
+3. The uploading client takes the SHA2-256 hash of the path component,
+   producing the *discovery key*.
+
+   For the above example, the discovery key is
+   ``af030c06750716b1b35852298fe852b90def13dcbd012a5fe5148470f1206bfc``. [#fn-hash]_
```

Comment:
This footnote links to a Unix-only command. Perhaps inline it and/or use a Python stanza?

Reply:
Thanks! Yeah, a Python equivalent makes sense, will do.